diff --git a/build/core/Makefile b/build/core/Makefile
--- a/build/core/Makefile
+++ b/build/core/Makefile
@@ -1315,8 +1315,6 @@ $(INTERNAL_OTA_PACKAGE_TARGET): $(BUILT_TARGET_FILES_PACKAGE) $(DISTTOOLS)
   $(OTA_FROM_TARGET_SCRIPT) -v \
      -p $(HOST_OUT) \
      -k $(KEY_CERT_PAIR) \
-     --no_assert=$(skip_assert) \
-     --persist=$(persist) \
      --override_device=$(override_device) \
      $(BUILT_TARGET_FILES_PACKAGE) $@
