diff --git a/frameworks/base/core/java/android/hardware/Camera.java b/frameworks/base/core/java/android/hardware/Camera.java
--- a/frameworks/base/core/java/android/hardware/Camera.java
+++ b/frameworks/base/core/java/android/hardware/Camera.java
@@ -3842,4 +3842,6 @@ private boolean same(String s1, String s2) {
             return false;
         }
     };
+    // Allwinner platform - set video screen
+    public static native int setCameraScreen(int screen);
 }
diff --git a/packages/apps/Camera/src/com/android/camera/FocusOverlayManager.java b/packages/apps/Camera/src/com/android/camera/FocusOverlayManager.java
--- a/packages/apps/Camera/src/com/android/camera/FocusOverlayManager.java
+++ b/packages/apps/Camera/src/com/android/camera/FocusOverlayManager.java
@@ -405,8 +405,8 @@ public void onCameraReleased() {
 
     private void autoFocus() {
         Log.v(TAG, "Start autofocus.");
-        mListener.autoFocus();
-        mState = STATE_FOCUSING;
+        //mListener.autoFocus();
+        mState = STATE_SUCCESS;
         // Pause the face view because the driver will keep sending face
         // callbacks after the focus completes.
         if (mFaceView != null) mFaceView.pause();
