diff --git a/packages/apps/Camera/src/com/android/camera/FocusOverlayManager.java b/packages/apps/Camera/src/com/android/camera/FocusOverlayManager.java
--- a/packages/apps/Camera/src/com/android/camera/FocusOverlayManager.java
+++ b/packages/apps/Camera/src/com/android/camera/FocusOverlayManager.java
@@ -406,7 +406,7 @@ public void onCameraReleased() {
     private void autoFocus() {
         Log.v(TAG, "Start autofocus.");
         mListener.autoFocus();
-        mState = STATE_FOCUSING;
+        mState = STATE_SUCCESS;
         // Pause the face view because the driver will keep sending face
         // callbacks after the focus completes.
         if (mFaceView != null) mFaceView.pause();
@@ -405,7 +405,7 @@ public void onCameraReleased() {
 
     private void autoFocus() {
         Log.v(TAG, "Start autofocus.");
-        mListener.autoFocus();
+        //mListener.autoFocus();
         mState = STATE_SUCCESS;
         // Pause the face view because the driver will keep sending face
         // callbacks after the focus completes.
