commit 91c51d72928816b5f0fdfcd0bf051bbfcf420f77
Author: fire855 <thefire855@gmail.com>
Date:   Fri Feb 20 23:38:43 2015 +0100

    Arnova 10c G3 Patches
    
    Change-Id: I2e47cfe8a76a0caf6bde85187dfb7692fa489699

diff --git a/BoardConfig.mk b/BoardConfig.mk
index 8278652..afb6f49 100644
--- a/BoardConfig.mk
+++ b/BoardConfig.mk
@@ -87,7 +87,7 @@ BOARD_CUSTOM_RECOVERY_KEYMAPPING := ../../device/softwinner/907/recovery/recover
 BOARD_UMS_LUNFILE := "/sys/class/android_usb/android0/f_mass_storage/lun/file"
 BOARD_UMS_2ND_LUNFILE := "/sys/class/android_usb/android0/f_mass_storage/lun1/file"
 RECOVERY_FSTAB_VERSION := 2
-BOARD_RECOVERY_SWIPE := true
+BOARD_HAS_NO_SELECT_BUTTON := true
 TARGET_RECOVERY_FSTAB := device/softwinner/907/rootdir/fstab.sun4i
 #TARGET_RECOVERY_PIXEL_FORMAT := "BGRA_8888"
 #TARGET_RECOVERY_UI_LIB := librecovery_ui_crane_evb
diff --git a/prebuilt/usr/keylayout/sun4i-keyboard.kl b/prebuilt/usr/keylayout/sun4i-keyboard.kl
old mode 100644
new mode 100755
index c9a0838..596cdb1
--- a/prebuilt/usr/keylayout/sun4i-keyboard.kl
+++ b/prebuilt/usr/keylayout/sun4i-keyboard.kl
@@ -1,4 +1,5 @@
 key 217   HOME              WAKE
-key 115	  VOLUME_UP
-key 114	  VOLUME_DOWN
-
+key 1     BACK              WAKE
+key 115	  MENU         	    WAKE_DROPPED
+key 102   VOLUME_UP
+key 28	  VOLUME_DOWN
diff --git a/rootdir/fstab.sun4i b/rootdir/fstab.sun4i
index a0a1873..e68e40c 100644
--- a/rootdir/fstab.sun4i
+++ b/rootdir/fstab.sun4i
@@ -7,7 +7,7 @@
 /dev/block/nande       /data          ext4     rw,noatime,nosuid,nodev,barrier=0,journal_checksum,noauto_da_alloc    wait
 /dev/block/nandh       /cache         ext4     rw,noatime,nosuid,nodev,barrier=0,journal_checksum,noauto_da_alloc    wait
 
-/devices/virtual/block/nand             /storage/sdcard0        vfat    rw,relatime,fmask=0000,dmask=0000    wait,check,noemulatedsd,voldmanaged=sdcard0:9
+/devices/virtual/block/nand             /storage/sdcard0        vfat    rw,relatime,fmask=0000,dmask=0000    wait,check,noemulatedsd,voldmanaged=sdcard0:10
 /devices/platform/sunxi-mmc.0/mmc_host  /storage/sdcard1        vfat    rw,relatime,fmask=0000,dmask=0000    wait,check,voldmanaged=sdcard1:auto
 /devices/platform/sunxi-mmc.1/mmc_host  /storage/sdcard1        vfat    rw,relatime,fmask=0000,dmask=0000    wait,check,voldmanaged=sdcard1:auto
 /devices/platform/sw-ehci.1             /storage/usbdisk0       vfat    rw,relatime,fmask=0000,dmask=0000    wait,check,voldmanaged=usbdisk0:auto
