commit 134f5a3ee53a207cec48e30ab8c73254f52b931e
Author: fire855 <thefire855@gmail.com>
Date:   Sat Feb 21 00:13:08 2015 +0100

    Arnova 10c G3 Patches
    
    Change-Id: I9628327ba0c87b4ff2bf38385c108efc4dfc1e11

diff --git a/907.mk b/907.mk
index e877146..941f68a 100644
--- a/907.mk
+++ b/907.mk
@@ -208,5 +208,5 @@ PRODUCT_AAPT_PREF_CONFIG := mdpi
 PRODUCT_NAME := full_907
 PRODUCT_DEVICE := 907
 
-TARGET_SCREEN_HEIGHT := 768
+TARGET_SCREEN_HEIGHT := 600
 TARGET_SCREEN_WIDTH := 1024
diff --git a/BoardConfig.mk b/BoardConfig.mk
index 8278652..afb6f49 100644
--- a/BoardConfig.mk
+++ b/BoardConfig.mk
@@ -87,7 +87,7 @@ BOARD_CUSTOM_RECOVERY_KEYMAPPING := ../../device/softwinner/907/recovery/recover
 BOARD_UMS_LUNFILE := "/sys/class/android_usb/android0/f_mass_storage/lun/file"
 BOARD_UMS_2ND_LUNFILE := "/sys/class/android_usb/android0/f_mass_storage/lun1/file"
 RECOVERY_FSTAB_VERSION := 2
-BOARD_RECOVERY_SWIPE := true
+BOARD_HAS_NO_SELECT_BUTTON := true
 TARGET_RECOVERY_FSTAB := device/softwinner/907/rootdir/fstab.sun4i
 #TARGET_RECOVERY_PIXEL_FORMAT := "BGRA_8888"
 #TARGET_RECOVERY_UI_LIB := librecovery_ui_crane_evb
diff --git a/prebuilt/usr/keylayout/sun4i-keyboard.kl b/prebuilt/usr/keylayout/sun4i-keyboard.kl
old mode 100644
new mode 100755
index c9a0838..596cdb1
--- a/prebuilt/usr/keylayout/sun4i-keyboard.kl
+++ b/prebuilt/usr/keylayout/sun4i-keyboard.kl
@@ -1,4 +1,5 @@
 key 217   HOME              WAKE
-key 115	  VOLUME_UP
-key 114	  VOLUME_DOWN
-
+key 1     BACK              WAKE
+key 115	  MENU         	    WAKE_DROPPED
+key 102   VOLUME_UP
+key 28	  VOLUME_DOWN
diff --git a/rootdir/fstab.sun4i b/rootdir/fstab.sun4i
index a0a1873..e68e40c 100644
--- a/rootdir/fstab.sun4i
+++ b/rootdir/fstab.sun4i
@@ -7,7 +7,7 @@
 /dev/block/nande       /data          ext4     rw,noatime,nosuid,nodev,barrier=0,journal_checksum,noauto_da_alloc    wait
 /dev/block/nandh       /cache         ext4     rw,noatime,nosuid,nodev,barrier=0,journal_checksum,noauto_da_alloc    wait
 
-/devices/virtual/block/nand             /storage/sdcard0        vfat    rw,relatime,fmask=0000,dmask=0000    wait,check,noemulatedsd,voldmanaged=sdcard0:9
+/devices/virtual/block/nand             /storage/sdcard0        vfat    rw,relatime,fmask=0000,dmask=0000    wait,check,noemulatedsd,voldmanaged=sdcard0:10
 /devices/platform/sunxi-mmc.0/mmc_host  /storage/sdcard1        vfat    rw,relatime,fmask=0000,dmask=0000    wait,check,voldmanaged=sdcard1:auto
 /devices/platform/sunxi-mmc.1/mmc_host  /storage/sdcard1        vfat    rw,relatime,fmask=0000,dmask=0000    wait,check,voldmanaged=sdcard1:auto
 /devices/platform/sw-ehci.1             /storage/usbdisk0       vfat    rw,relatime,fmask=0000,dmask=0000    wait,check,voldmanaged=usbdisk0:auto
diff --git a/system.prop b/system.prop
index 21a82e6..3ba6ba8 100644
--- a/system.prop
+++ b/system.prop
@@ -47,5 +47,5 @@ ro.device.cpu=1.0 GHz AllWinner A10
 ro.device.gpu=ARM Mali 400
 ro.device.rear_cam=2 MP
 ro.device.front_cam=0.3 MP
-ro.device.screen_res=1024x768
+ro.device.screen_res=1024x600
 ro.chip_type=a10
