From 5c7689d9c069debd9d50fa16e40dc7394a6643b7
From: ugers <yugers@gmail.com>
Date: Thu, 26 Feb 2015 21:37:46 +0300
Subject: [PATCH] system_core
  
Change-Id: Ifdc9422bc28b1f4ed06312ff389d50a39d3c7bbb

diff --git a/include/system/camera.h b/include/system/camera.h
index 9cc8286..e5e3b1e 100644
--- a/include/system/camera.h
+++ b/include/system/camera.h
@@ -87,22 +87,11 @@ enum {
     CAMERA_MSG_PREVIEW_METADATA = 0x0400, // dataCallback
     // Notify on autofocus start and stop. This is useful in continuous
     // autofocus - FOCUS_MODE_CONTINUOUS_VIDEO and FOCUS_MODE_CONTINUOUS_PICTURE.
-#if defined(QCOM_ICS_COMPAT) && defined(QCOM_HARDWARE)
-    CAMERA_MSG_STATS_DATA       = 0x800,
-    CAMERA_MSG_FOCUS_MOVE = 0x1000,       // notifyCallback
-#elif defined(OMAP_ICS_CAMERA) && defined(OMAP_ENHANCEMENT)
-    CAMERA_MSG_COMPRESSED_BURST_IMAGE = 0x0800, //dataCallback
-    CAMERA_MSG_RAW_BURST = 0x1000,        // dataCallback
-#else
     CAMERA_MSG_FOCUS_MOVE = 0x0800,       // notifyCallback
-#ifdef QCOM_HARDWARE
-    CAMERA_MSG_STATS_DATA       = 0x1000,
-    CAMERA_MSG_META_DATA = 0x2000,
-#elif defined(OMAP_ENHANCEMENT) && defined(OMAP_ENHANCEMENT_BURST_CAPTURE)
-    CAMERA_MSG_COMPRESSED_BURST_IMAGE = 0x1000, // dataCallback
-    CAMERA_MSG_RAW_BURST = 0x2000,        // dataCallback
-#endif
-#endif
+    CAMERA_MSG_CONTINUOUSSNAP= 0x1000,    //notifyCallback for continuous snap by fuqiang
+    CAMERA_MSG_SNAP= 0x2000,    //notifyCallback of setting camera idle  for single snap by fuqiang
+    CAMERA_MSG_SNAP_THUMB= 0x4000,    //notifyCallback of saving thumb for single snap by fuqiang
+    CAMERA_MSG_SNAP_FD= 0x8000,    //notifyCallback of requesting fd for single and continuoussnap by fuqiang
     CAMERA_MSG_ALL_MSGS = 0xFFFF
 };
 
